### ISSUE TITLE
Crawcin(), Cconin(), Cnecin(), ... return incorrect (?) scan codes for special keys

### ISSUE THREAD

**Issue by mikrosk** [freemint/freemint#281]
I don't know if I'm overlooking something obvious but on a first sight it seems that FreeMiNT is totally broken in this regard.

For instance Crawcin() + left arrow: it should return something like `0x004b0000` like TOS does. However it returns `0x0001001b`, i.e. like if I pressed ESC key. I get the exact same value for right arrow, F1, ... but normal letters (A, B, C, ...) return expected results.

TOS (4.04) always returns expected values.

Tested with all three functions.

I guess TosWin2 doesn't play a role here?

**Comment by lpgb** [comment]
Tested cconin() with kernel 1.16 and get 0x004B0000 for left arrow and 0x004D0000 for right arrow. I had TosWin2 running during the test.

**Comment by th-otto** [comment]
On Freitag, 7. Oktober 2022 23:54:11 CEST Miro Kropáček wrote:

I think that is done by toswin, not the kernel. When pressing cursor-up, you 
get three chars: esc, '[', 'D'. Similar for other function keys.

**Comment by mikrosk** [comment]
@th-otto is there a way to prevent this behaviour? I mean, what a poor, TOS-only .TOS/.TTP has to know about FreeMiNT / terminal emulator if it is only interested in scan codes?

@lpgb that's interesting, I tried 1.16.1 + recent TosWin2 and I get the same behaviour. I presume you pressed the keys within the console window?

**Comment by mikrosk** [comment]
Ehem, ok, I have an update. It seems that this behaviour is related to the terminal setting -- `tw100` leads to the broken behaviour while `tw52` leads to expected behaviour.

Shouldn't this be somehow handled by `Pdomain` or such? TOS applications shouldn't care about terminal settings at all.

**Comment by lpgb** [comment]
> @lpgb that's interesting, I tried 1.16.1 + recent TosWin2 and I get the same behaviour. I presume you pressed the keys within the console window?

Had TosWin2 running and the console open, but no I ran the test as a PRG.

**Comment by mikrosk** [comment]
That explains it then, it is the .TOS file extension which makes it behave this way.

**Comment by mikrosk** [comment]
I was curious how @pmandin did it in SDL, as his PmDoom is a .ttp and regardless if I ran it with `SDL_VIDEODRIVER=gem` or `SDL_VIDEODRIVER=xbios` from TosWin2, the cursor keys worked.

Well, as it turned out, it uses own IKBD handler by default. :) As soon as I set `SDL_ATARI_EVENTSDRIVER` to `gemdos` or `bios`, no fun, got only the escape codes with `SDL_VIDEODRIVER=xbios`. With `SDL_VIDEODRIVER=gem` it's clear, it is the top window, not TosWin2, which receives key strokes, so the events driver works with `ikbd`, `gemdos` as well as `bios`.

So yeah, I did the right thing and went down with separate AES+`evnt_keybd()` and GEMDOS handling, works nicely.

**Comment by pmandin** [comment]
When I wrote gemdos and bios keyboard support for SDL, I only tested with TOS, and MiNT with tty console (text mode, no AES). And with MiNT with AES, I only tested the AES/VDI backend of course, I suppose that's why I never encountered this issue.

However I remember when writing the cleancode demo programs there was something strange when ran from Toswin, but I never dug the issue, thinking it was Mint, or the AES, or something else.
